### PR TITLE
Implement scrollbar-width CSS property in WebKit2

### PR DESCRIPTION
**Pull request by lukewarlow**
#### 7a7bb7a3e5e23541e6ca556920b8359cf9c7b785
<pre>
Implement scrollbar-width CSS property in WebKit2
<a href="https://bugs.webkit.org/show_bug.cgi?id=256999">https://bugs.webkit.org/show_bug.cgi?id=256999</a>

Reviewed by Simon Fraser.

This change implements the basics of the scrollbar-width CSS property.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-keywords-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::createScrollbar):
(WebCore::LocalFrameView::scrollbarWidthStyle const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::scrollbarWidthStyle const):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::Scrollbar):
(WebCore::Scrollbar::isHiddenByStyle const):
* Source/WebCore/platform/Scrollbar.h:
(WebCore::Scrollbar::isHiddenByStyle const): Deleted.
* Source/WebCore/platform/ScrollbarTheme.h:
(WebCore::ScrollbarTheme::scrollbarThickness):
* Source/WebCore/platform/ScrollbarThemeComposite.cpp:
(WebCore::ScrollbarThemeComposite::minimumThumbLength):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
(WebCore::ScrollbarThemeAdwaita::scrollbarThickness):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.h:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
(WebCore::ScrollbarThemeGtk::scrollbarThickness):
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.mm:
(WebCore::ScrollbarThemeIOS::scrollbarThickness):
* Source/WebCore/platform/mac/ScrollbarThemeMac.h:
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::scrollbarThickness):
(WebCore::ScrollbarThemeMac::backButtonRect):
(WebCore::ScrollbarThemeMac::forwardButtonRect):
(WebCore::ScrollbarThemeMac::trackRect):
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
* Source/WebCore/platform/mock/ScrollbarThemeMock.cpp:
(WebCore::ScrollbarThemeMock::scrollbarThickness):
* Source/WebCore/platform/mock/ScrollbarThemeMock.h:
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.cpp:
(WebCore::ScrollbarThemePlayStation::scrollbarThickness):
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.h:
* Source/WebCore/platform/win/ScrollbarThemeWin.cpp:
(WebCore::ScrollbarThemeWin::scrollbarThickness):
* Source/WebCore/platform/win/ScrollbarThemeWin.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbar):
(WebCore::RenderLayerScrollableArea::scrollbarWidthStyle const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderScrollbar.h:
* Source/WebCore/rendering/RenderScrollbarTheme.h:

Canonical link: <a href="https://commits.webkit.org/264593@main">https://commits.webkit.org/264593@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2334d700df4cf0ecc424da09649af633a41e1355

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8303 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9670 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8122 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8214 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10987 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8162 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9249 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7321 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9790 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6550 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14918 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7673 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7459 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10821 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7930 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6451 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7244 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1935 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11452 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7671 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/LocalFrameView.cpp
Comment (smfr) [inline Source/WebCore/page/LocalFrameView.cpp]: `auto scrollbarSize = ` (better than just "size" because size is normally a thing with height and width).

File: Source/WebCore/page/LocalFrameView.cpp
Comment (smfr) [inline Source/WebCore/page/LocalFrameView.cpp]: Is `documentElement` the right thing to check for the style? How about `scrollingElement`? Does `scrollbar-width` propagate from the body to the document like `overflow` does?

File: Source/WebCore/platform/ScrollbarTheme.h
Comment (smfr) [inline Source/WebCore/platform/ScrollbarTheme.h]: The bool param needs a name (commented out, since the implementation is empty). Or it might be cleaner to just pass the ScrollbarWidth enum.

File: Source/WebCore/rendering/RenderLayerScrollableArea.cpp
Comment (smfr) [inline Source/WebCore/rendering/RenderLayerScrollableArea.cpp]: `auto scrollbarSize = `

File: Source/WebCore/platform/ScrollableArea.h
Comment (smfr) [inline Source/WebCore/platform/ScrollableArea.h]: Slightly uneasy with the name (it'd expect it to return an `int`). Maybe `scrollbarWidthStyle()`?

**Review by smfr — CHANGES_REQUESTED** [review]
Let's do an update passing ScrollbarWidth instead of the bool, and checking the style propagation question.

File: Source/WebCore/platform/ScrollbarTheme.h
Comment (lukewarlow) [reply Source/WebCore/platform/ScrollbarTheme.h:54]: Given that auto and thin are accounted for in the controls size, none could maybe be added in that?

 Or alternatively we could just pass scrollbar width and not control size but then we'd need to do the mapping in multiple places?

Happy to just pass down the ScrollbarWidth value though.

File: Source/WebCore/page/LocalFrameView.cpp
Comment (lukewarlow) [reply Source/WebCore/page/LocalFrameView.cpp]: This reads to me like the answer is no it doesn't propagate and as such I think you're right it should be scrollingElement not document element.

https://drafts.csswg.org/css-scrollbars/#scrollbar-width:~:text=is%20not%20propagated%20to%20the%20viewport.

image

File: Source/WebCore/platform/ScrollbarTheme.h
Comment (lukewarlow) [reply Source/WebCore/platform/ScrollbarTheme.h]: That makes sense I will replace it with the ScrollbarWidth enum. A follow up change could probably remove the ScrollbarControlSize enum entirely as ScrollbarWidth should be able to replace it entirely

**Comment by lukewarlow** [comment]
There's a lot of failing GTK tests but from what I can work out the only "new" ones are webgl conformance tests which I've seen fail on a few branches. So I think that failure is unrelated to my changes.

File: Source/WebCore/platform/ScrollbarTheme.h
Comment (lukewarlow) [reply Source/WebCore/platform/ScrollbarTheme.h]: @smfr should I make the bug to refactor ScrollbarControlSize away now or should I wait till this is merged and then do it?

**Comment by lukewarlow** [comment]
https://github.com/WebKit/standards-positions/issues/133 - given this property will be mostly implemented once this PR is merged it might be worth chucking down a position on this issue.

File: Source/WebCore/platform/ScrollbarTheme.h
Comment (smfr) [reply Source/WebCore/platform/ScrollbarTheme.h]: You can do it separately.

**Review by smfr — APPROVED** [review]

File: Source/WebCore/page/LocalFrameView.cpp
Comment (smfr) [reply Source/WebCore/page/LocalFrameView.cpp]: Make sure there at WPT tests that test that.

Comment (nt1m) [reply Source/WebCore/page/LocalFrameView.cpp]: @lukewarlow ^ can you please do this as a followup?

**Pull request merged**
